src/repository/sessions.rs

```diff
@@ -0,0 +1,96 @@
+use axum::async_trait;
+use sqlx::PgPool;
+use tracing::error;
+
+use crate::domain::sessions::Session;
+
+pub enum SessionGetError {
+    Missing,
+    Unknown
+}
+
+pub enum SessionDeleteError {
+    Unknown
+}
+
+pub enum SessionInsertError {
+    Unknown
+}
+
+#[async_trait]
+pub trait SessionRepository {
+    async fn insert(&self, session: &Session) -> Result<(), SessionInsertError>;
+    async fn get(&self, id: &str) -> Result<Session, SessionGetError>;
+    async fn delete(&self, id: &str) -> Result<(), SessionDeleteError>;
+}
+
+#[derive(Debug, Clone)]
+pub struct PgSessionRepository {
+    pub pool: PgPool
+}
+
+impl PgSessionRepository {
+    pub fn new(pool: PgPool) -> Self {
+        Self { pool }
+    }
+}
+
+#[async_trait]
+impl SessionRepository for PgSessionRepository {
+    #[tracing::instrument(skip(self))]
+    async fn insert(&self, session: &Session) -> Result<(), SessionInsertError> {
+        match sqlx::query("INSERT INTO sessions (id, user_id, expires) VALUES ($1, $2, $3)")
+            .bind(&session.id)
+            .bind(&session.user.id)
+            .bind(&session.expires)
+            .execute(&self.pool)
+            .await {
+                Ok(_) => Ok(()),
+                Err(err) => {
+                    error!(%err);
+                    Err(SessionInsertError::Unknown)
+                }
+            }
+    }
+
+    #[tracing::instrument(skip(self))]
+    async fn get(&self, id: &str) -> Result<Session, SessionGetError> {
+        let session = sqlx::query_as::<_, Session>("
+            SELECT
+                sessions.id AS session_id, 
+                sessions.user_id AS user_id, 
+                sessions.expires as expires, 
+                users.email AS email, 
+                users.password_hash AS password_hash
+            FROM sessions, users
+            WHERE sessions.id = $1
+        ")
```

Comment:
Missing `JOIN`

Reply:
The comma is INNER JIOIN

Reply:
It doesn't seem to work like that. Local test:
```
INSERT INTO users(id, email, password_hash) VALUES (1, 'a', 'a'), (2, 'b', 'b');
INSERT INTO sessions(id, user_id, expires) VALUES (1, 1, 1), (2, 2, 2);
SELECT count(*) FROM sessions, users; /* returns 4, should be 2 */
```

Reply:
Oh I forgot the condition